src/ethereum/starknet/state_root.rs

```diff
@@ -0,0 +1,294 @@
+        //
+        // Allows us to perform a binary search of the query range space.
+        let mut stride_cap = None;
+
+        loop {
+            let to_block = from_block.saturating_add(self.stride);
+
+            let mut logs = match self.get_logs(from_block, to_block).await {
+                Ok(logs) => logs,
+                Err(GetLogsError::QueryLimit) => {
+                    // Infura result cap error. We need to reduce our query range.
+                    stride_cap = Some(self.stride);
+                    self.stride = (self.stride / 2).max(1);
+
+                    continue;
+                }
+                Err(GetLogsError::Other(other)) => return Err(GetRootError::Other(other)),
+            };
+
+            // Check for reorgs. Only required if there was a last known update to validate.
+            if let Some(last) = self.last.as_ref() {
+                // We queried for logs starting from the same block as last known. We need to account
+                // for logs that occurred in the same block and transaction, but with a smaller log index.
+                //
+                // If the last known log is not in the set then we have a reorg event.
+                loop {
+                    match logs.pop_front() {
+                        // Log from the same origin point, but smaller log index.
+                        Some(log)
+                            if log.origin == last.origin && log.log_index < last.log_index =>
+                        {
+                            continue
+                        }
+                        // Log found. Yay!
+                        Some(log) if &log == last => break,
+                        _ => return Err(GetRootError::Reorg),
+                    }
+                }
+            }
+
+            // At this point we have validated and removed the last known log.
+            // If there are no new logs then either:
+            //  1. there are no new logs on L1, or
+            //  2. we need to increase our stride
+            if logs.is_empty() {
+                let latest_on_chain = self.websocket.eth().block_number().await?.as_u64();
+
+                if to_block <= latest_on_chain {
+                    // Increase stride. Account for stride cap (binary search style)
+                    // so that we don't run into a cycle of:
+                    //  1. No logs, increase stride
+                    //  2. Infura result cap, reduce stride
+                    match stride_cap {
+                        Some(max) => self.stride = (self.stride.saturating_add(max + 1)) / 2,
+                        None => self.stride = self.stride.saturating_mul(2),
+                    }
+                } else {
+                    // No new logs on L1, sleep and try again..
+                    const SLEEP: Duration = Duration::from_secs(5);
+                    tokio::time::sleep(SLEEP).await;
+                }
```

Comment:
So in the else case why is there a another run with possibly the same `from_block` and `to_block`? I assume that those are unchanged in the `!log.is_empty()` case as there doesn't seem to be any path which would increase `from_block` and thus `to_block` would also remain the same. Or maybe I'm not seeing some part.

Would it not be good to just return the found logs?

Reply:
The premise is that if we have gotten to this point, that there are no new logs to output. So we try again a bit later and hope that there is a new block which contains new logs. If there is no new block with new logs we will end up here again. And repeat.

The `from_block` has to stay the same, in order to check for reorgs (every `get_logs` call must include the last block we knew of).

The only thing that would eventually change is that a new block arrives.

Reply:
We are effectively caught up to L1 (in sync) at that point. It may be smarter to move to using a log subscription at that point -- so that we get notified of new logs instead of polling. But that has its downsides as well.

Reply:
I am not sure I follow. I was talking about the else case, where `!logs.is_empty()`. There's currently no such in the code (on the commented line).

Should that if be turned into an assertion then as `assert!(logs.is_empty());`? But even with that I'm still not seeing when will this function get out of the loop.

Reply:
It will stay inside the loop until there is a change in data from L1. Note that the if has a [small bug (comment)](https://github.com/eqlabs/pathfinder/pull/34/files#r756883568).

Imagine the following scenario. We have consumed all state update logs that are available on L1 currently. The last state update log that we had was at block `latest_on_chain - 5`. We have `stride = 10`.

Now when we call `get_logs` we only ever get a single log back. Which is the last log update we already handed out.

```rust
// short pseudo-code for what is happening:
let logs = get_logs(from = latest_on_chain - 5, to = latest_on_chain - 5 + 10);
assert_eq!(logs.len(), 1);
assert_eq!(logs[0], self.last);
```

So we keep looping until L1 adds another block and that block contains another new log. At which point we will exit the loop. Or, alternatively many blocks get added without any new logs at which point `stride` will be increased. Or, finally a reorg happens which invalidates our `self.last`.

Reply:
Effectively its polling L1 for new data. If there is no new data, sleep and try again.

Unless I have a bug, this does spell out that I need better comments around this then 😢

Reply:
No it's just me misreading the indentation line. I read this as:

```
loop {

}

return Ok(logs);
```

While it was:

```
loop {
  // dostuff

  return Ok(logs);
}
```

So this is ready to be resolved, apologies.